### PR TITLE
Add metric for start/end task run

### PR DESCRIPTION
**Pull request by Acehaidrey**
The purpose of this metric is to have a value for when a stat begins and when it ends to set up different rules on these stats in the prometheus side/statsd side etc. This allows for easily building reports as well and is a minor change to enable teams, while also allowing users to black list these stats if they do not want it (black listing changes already committed with community team changes).

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by Acehaidrey** [comment]
@jhtimmins if you don't mind when you get a chance to look at some of this and share your thoughts as well

File: airflow/models/taskinstance.py
Comment (jhtimmins) [inline airflow/models/taskinstance.py]: To make these names as meaningful as possible, `starts` and `finishes` (or similar) might be more clear, so it's clear that it's referring to a counter and not a timestamp.

https://airflow.apache.org/docs/stable/metrics.html#counters
* `ti_starts_<dag_id>_<task_id>`
* `ti_finishes_<dag_id>_<task_id>_<state>`

Comment (jhtimmins) [reply airflow/models/taskinstance.py]: Just wondering, does it make more sense to put it here rather than after the success callbacks on lines 1087-1093, inside a `finally` block?

**Review by jhtimmins — CHANGES_REQUESTED** [review]
The only changes I'm requesting are the stats name clarification and testing for the `start` counter.

The other comment is a question, not a suggestion, necessarily.

I like these suggestions though. I think this update makes sense.

File: airflow/models/taskinstance.py
Comment (Acehaidrey) [reply airflow/models/taskinstance.py]: change it to finished . and moved to a finally so it is better this way
I do the separation on the `.` because can do different grouping aggregates on statsd/prometheus side vs using the `_`

**Comment by Acehaidrey** [comment]
I made the following changes:
* changed the word complete to finish
* added test to verify ti.start is also in the calls for the function
* updated documentation reflecting those changes

I do the separation on the `.` because can do different grouping aggregates on statsd/prometheus side vs using the _

**Comment by Acehaidrey** [comment]
@jhtimmins when you get a chance

**Comment by Acehaidrey** [comment]
@jhtimmins sorry for pestering but would love to get these in and close out

**Comment by jhtimmins** [comment]
@Acehaidrey Apologies, didn't realize I was a blocker on these. I will take a look shortly. 

I don't have commit privileges however, so a committer will need to review/approve as well.

**Comment by Acehaidrey** [comment]
Not a problem! Just appreciate your help in signing off then I will ping @ashb if can take a look too. You're awesome thanks again

File: airflow/models/taskinstance.py
Comment (ashb) [inline airflow/models/taskinstance.py]:
```suggestion
            Stats.incr('ti.finish.{}.{}.{}'.format(task.dag_id, task.task_id, self.state))
```

(I don't know why we include those two default params elsewhere.)

Comment (Acehaidrey) [reply airflow/models/taskinstance.py]: Thank you @ashb can you make the same recommendation for the start task so I can just merged in the suggestion here too

File: airflow/models/taskinstance.py
Comment (Acehaidrey) [inline airflow/models/taskinstance.py]: Stats.incr('ti.start.{}.{}'.format(task.dag_id, task.task_id))

**Review by Acehaidrey — COMMENTED** [review]
@ashb mind taking another look? I have made the updates asked

File: airflow/models/taskinstance.py
Comment (Acehaidrey) [reply airflow/models/taskinstance.py]: @ashb mind taking a look at thiss so we can close this oout. did the changes asked

**Comment by Acehaidrey** [comment]
@ashb @jhtimmins mind helping me close this one out?

**Comment by jhtimmins** [comment]
Hi @Acehaidrey, sorry been heads down with 3.8 stuff. Happy to take a look tomorrow. thanks for pinging me.

**Comment by Acehaidrey** [comment]
not a problem @jhtimmins I know you are all busy! please let me know if I can help ever! would love to be closer to community :) but regarding this PR was there anything else, I did all your changes you asked, so want to make sure.

**Review by dimberman — APPROVED** [review]

**Comment by dimberman** [comment]
@Acehaidrey can you please fix the tests? Will merge when passing.

**Comment by Acehaidrey** [comment]
@dimberman yes I can rebase and try again, it seems the tests are failing to some connection issue with hive which are not related to this. I'm thinking its intermittent ?

**Comment by Acehaidrey** [comment]
hey @dimberman @ashb I am getting some exit code 137 meaning oom issues. that is why tests are failing. but wondering if you knew where to touch those containers to request more memory to make it pass?

**Comment by jhtimmins** [comment]
@Acehaidrey Not sure about increasing the memory, but several of us are actively working to reduce the image size to solve that issue. Hard to give an exact timeline, but can say that it's in progress.

**Comment by Acehaidrey** [comment]
hey @jhtimmins thanks for the update! That is okay with me. I just don't know how to make this test pass otherwise and @dimberman has asked for it

**Comment by Acehaidrey** [comment]
@dimberman mind chiming in when get a chance?

**Comment by dimberman** [comment]
Hi @Acehaidrey I'm currently fixing an issue where travis can't seem to run the k8s tests. I'm trying to avoid merging anything until that is fixed. I'll make sure to come back to this when that merges.

**Comment by Acehaidrey** [comment]
@dimberman  not a problem let me know if I can help and thanks for your reply~

**Comment by dimberman** [comment]
Hi @Acehaidrey. The tests are working now. Can you please rebase?

**Comment by Acehaidrey** [comment]
@dimberman done! it has all passed

**Comment by turbaszek** [comment]
Hi @Acehaidrey it seems that you have performed merge instead of rebase. Can you please do `git rebase apache/master` (assuming that your Apache remote is called apache)?
https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#id39

**Comment by Acehaidrey** [comment]
@turbaszek my apologies - check now ( once build is successful that is )

**Review by turbaszek — APPROVED** [review]

**Pull request merged**
